contracts/src/v0.8/llo-feeds/dev/FeeManager.sol

```diff
@@ -353,4 +331,113 @@ contract FeeManager is IFeeManager, ConfirmedOwner, TypeAndVersionInterface {
+      (, , , , , bytes memory quoteBytes) = abi.decode(
+        payload,
+        // reportContext, report, rs, ss, raw, quote
+        (bytes32[3], bytes, bytes32[], bytes32[], bytes32, bytes)
+      );
+
+      //decode the quote from the bytes
+      (quote) = abi.decode(quoteBytes, (Quote));
+    }
+
+    //decode the fee, it will always be native or LINK
+    return getFeeAndReward(msg.sender, report, quote);
+  }
+
+  function _handleFeesAndRewards(
+    address subscriber,
+    FeeAndReward[] memory feesAndRewards,
+    uint256 numberOfLinkFees,
+    uint256 numberOfNativeFees
+  ) internal {
+    IRewardManager.FeePayment[] memory linkFees = new IRewardManager.FeePayment[](numberOfLinkFees);
+    IRewardManager.FeePayment[] memory nativeFees = new IRewardManager.FeePayment[](numberOfNativeFees);
+
+    uint256 totalNativeFee;
+    uint256 totalNativeFeeLinkValue;
+
+    uint256 linkFeesIndex;
+    uint256 nativeFeesIndex;
+
+    uint256 totalNumberOfFees = numberOfLinkFees + numberOfNativeFees;
+    for (uint256 i; i < totalNumberOfFees; ++i) {
+      if (feesAndRewards[i].fee.assetAddress == i_linkAddress) {
+        linkFees[linkFeesIndex++] = IRewardManager.FeePayment(
+          feesAndRewards[i].configDigest,
+          uint192(feesAndRewards[i].reward.amount)
+        );
+      } else {
+        nativeFees[nativeFeesIndex++] = IRewardManager.FeePayment(
+          feesAndRewards[i].configDigest,
+          uint192(feesAndRewards[i].reward.amount)
+        );
+        totalNativeFee += feesAndRewards[i].fee.amount;
+        totalNativeFeeLinkValue += feesAndRewards[i].reward.amount;
+      }
+    }
+
+    //keep track of change in case of any over payment
+    uint256 change;
+
+    if (msg.value != 0) {
+      //there must be enough to cover the fee
+      if (totalNativeFee > msg.value) revert InvalidDeposit();
+
+      //wrap the amount required to pay the fee & approve as the subscriber paid in unwrapped native
+      IWERC20(i_nativeAddress).deposit{value: totalNativeFee}();
+
+      unchecked {
+        //msg.value is always >= to fee.amount
+        change = msg.value - totalNativeFee;
+      }
+    } else {
+      if (totalNativeFee != 0) {
+        //subscriber has paid in wrapped native, so transfer the native to this contract
+        IERC20(i_nativeAddress).transferFrom(subscriber, address(this), totalNativeFee);
+      }
+    }
+
+    if (linkFees.length != 0) {
+      i_rewardManager.onFeePaid(linkFees, subscriber);
+    }
+
+    if (nativeFees.length != 0) {
+      //distribute subsidised fees paid in Native
+      if (totalNativeFeeLinkValue > IERC20(i_linkAddress).balanceOf(address(this))) {
+        // If not enough LINK on this contract to forward for rewards, fire this event and
+        // call onFeePaid out-of-band to pay out rewards
+        emit InsufficientLink(nativeFees);
+      } else {
+        //distribute the fees
+        i_rewardManager.onFeePaid(nativeFees, address(this));
```

Comment:
This isn't right. `onFeePaid` can only be called with LINK and not with native token fees. I'll take a crack at fixing it so we can get contracts deployed.